lib/elixir/lib/code/normalizer.ex

```diff
@@ -0,0 +1,246 @@
+defmodule Code.Normalizer do
+  @moduledoc false
+
+  defguard is_literal(x)
+           when is_integer(x) or
+                  is_float(x) or
+                  is_binary(x) or
+                  is_atom(x)
+
+  @doc """
+  Wraps literals in the quoted expression to conform to the AST format expected
+  by the formatter.
+  """
+  def normalize(quoted, opts \\ []) do
+    line = Keyword.get(opts, :line, 1)
+
+    do_normalize(quoted, line: line)
+  end
+
+  # Skip already wrapped blocks
+  defp do_normalize({:__block__, _, [literal]} = quoted, _parent_meta) when is_literal(literal) do
+    quoted
+  end
+
+  defp do_normalize({:__block__, meta, args} = quoted, _parent_meta) do
+    if Keyword.has_key?(meta, :format) do
+      quoted
+    else
+      {:__block__, meta, Enum.map(args, &do_normalize(&1, meta))}
+    end
+  end
+
+  # Skip aliases so the module segment atoms don't get wrapped
+  defp do_normalize({:__aliases__, _, _} = quoted, _parent_meta) do
+    quoted
+  end
+
+  # Skip qualified tuples left hand side
+  defp do_normalize({:., _, [_, :{}]} = quoted, _parent_meta) do
+    quoted
+  end
+
+  # Don't normalize the `Access` atom in access syntax
+  defp do_normalize({:., _, [Access, :get]} = quoted, _parent_meta) do
+    quoted
+  end
+
+  # Only normalize the left side of the dot operator
+  # The right hand side is an atom in the AST but it's not an atom literal, so
+  # it should not be wrapped
+  defp do_normalize({:., meta, [left, right]}, _parent_meta) do
+    {:., meta, [do_normalize(left, meta), right]}
+  end
+
+  # A list of left to right arrows is not considered as a list literal, so it's
+  # not wrapped
+  defp do_normalize([{:->, _, _} | _] = quoted, parent_meta) do
+    Enum.map(quoted, &do_normalize(&1, parent_meta))
+  end
+
+  # left -> right
+  defp do_normalize({:->, meta, [left, right]}, _parent_meta) do
+    left = Enum.map(left, &do_normalize(&1, meta))
+    right = do_normalize(right, meta)
+    {:->, meta, [left, right]}
+  end
+
+  # Maps
+  defp do_normalize({:%{}, meta, args}, _parent_meta) do
+    args = Enum.map(args, &do_normalize(&1, meta))
+
+    args =
+      case args do
+        # Unwrap the right hand side if we're in an update syntax
```

Comment:
I would prefer to follow the same style as the remaining of code which is to avoid wrapping in the first place (instead of unwrapping later). That's because I can always understand the input (I know which shapes are valid AST) but I don't know all possible outputs for normalize. :) Would it be possible to change this?